### PR TITLE
Shuffle between shards

### PR DESCRIPTION
**Pull request by raduchis**
added functionality to reshuffle nodes between shards. The node is restarted and the mechanism for startInEpoch kicks in requesting the missing functionality.

- if the validator is shuffled out of one shard it will restart and move to another shard

### CODE REVIEW

File: cmd/node/main.go [hunk 1]
```diff
@@ -1115,9 +1138,48 @@ func startNode(ctx *cli.Context, log logger.Logger, version string) error {
 	err = networkComponents.NetMessenger.Close()
 	log.LogIfError(err)
 
+	handleAppClose(log, sig)
+
 	return nil
 }
 
+func handleAppClose(log logger.Logger, endProcessArgument endProcess.EndProcessArgument) {
+	log.Debug(fmt.Sprintf("Restarting because %s. Details: %s",
```
Comment (SebastianMarian) [inline cmd/node/main.go:1147]: Why not directly: log.Debug("Restarting", "reason", endProcessArgument.Reason, "details", endProcessArgument.Description)

File: cmd/node/main.go [hunk 2]
```diff
@@ -1115,9 +1138,48 @@ func startNode(ctx *cli.Context, log logger.Logger, version string) error {
...
+func newStartInEpoch(log logger.Logger) {
+	wd, err := os.Getwd()
+	if err != nil {
+		log.LogIfError(err)
+	}
+	nodeApp := os.Args[0]
+	args := os.Args
+	args = append(args, "-start-in-epoch")
+
+	log.Debug("Executable", "working dir", wd)
```
Comment (SebastianMarian) [inline cmd/node/main.go:1167]: Why not directly one line: log.Debug("Executable", "working dir", wd, "node app", nodeApp, "args", args) and remove lines 1168 and 1169

File: cmd/node/main.go [hunk 3]
```diff
@@ -1475,6 +1555,35 @@ func createNodesCoordinator(
 	return nodesCoordinator, nil
 }
 
+func serializableValidatorsToValidators(nodeRegistryValidators map[string][]*sharding.SerializableValidator) (map[uint32][]sharding.Validator, error) {
+	validators := make(map[uint32][]sharding.Validator)
+	for shardId, shardValidators := range nodeRegistryValidators {
+		newValidators, err2 := serializableShardValidatorListToValidatorList(shardValidators)
```
Comment (SebastianMarian) [inline cmd/node/main.go:1561]: Why not err instead err2?

File: core/constants.go [hunk 4]
```diff
@@ -338,3 +338,12 @@ const (
...
 	BuiltInFunctionSaveKeyValue = "SaveKeyValue"
 )
+
+// ShuffledOut signals that a restart is pending because the node was shuffled out
+const ShuffledOut = "shuffledOut"
+
+// MaxRetriesToCreateDB represents the maximum number of times to try to create DB if it failed
+const MaxRetriesToCreateDB = 10
+
+// SleepTimeBetweenCreateDBRetries representes the number of seconds to sleep between DB creates
```
Comment (SebastianMarian) [inline core/constants.go:348]: represents

File: scripts/testnet/variables.sh [hunk 5]
```diff
@@ -49,15 +49,15 @@ export TX_SIGN_FORMAT="bech32"
 export BLOCK_SIGN_FORMAT="hex"
 
 # Shard structure
-export SHARDCOUNT=2
-export SHARD_VALIDATORCOUNT=3
+export SHARDCOUNT=1
```
Comment (SebastianMarian) [inline scripts/testnet/variables.sh:52]: Do we want to modify the default values from this script?

File: sharding/indexHashedNodesCoordinator.go [hunk 6]
```diff
@@ -115,6 +115,14 @@ func NewIndexHashedNodesCoordinator(arguments ArgNodesCoordinator) (*indexHashed
 		log.Error("saving initial nodes coordinator config failed",
 			"error", err.Error())
 	}
+	log.Info("new nodes config is set for epoch", "epoch", arguments.Epoch)
+	currentNodesConfig := ihgs.nodesConfig[arguments.Epoch]
+	displayNodesConfiguration(
```
Comment (SebastianMarian) [inline sharding/indexHashedNodesCoordinator.go:120]: This print should be done only if currentNodesConfig is not nil, otherwise it will panicked here

File: sharding/indexHashedNodesCoordinator.go [hunk 7]
```diff
@@ -201,18 +210,18 @@ func (ihgs *indexHashedNodesCoordinator) setNodesPerShards(
 	nodesConfig.eligibleMap = eligible
 	nodesConfig.waitingMap = waiting
 	nodesConfig.publicKeyToValidatorMap = ihgs.createPublicKeyToValidatorMap(eligible, waiting)
+	log.Info("from nodesCoordinator")
```
Comment (SebastianMarian) [inline sharding/indexHashedNodesCoordinator.go:213]: Do we need this print?

File: sharding/indexHashedNodesCoordinator.go [hunk 8]
```diff
@@ -201,18 +210,18 @@ func (ihgs *indexHashedNodesCoordinator) setNodesPerShards(
...
 	nodesConfig.shardID = ihgs.computeShardForSelfPublicKey(nodesConfig)
 	nodesConfig.selectors, err = ihgs.createSelectors(nodesConfig)
 	if err != nil {
 		return err
 	}
 
-	shardIDForSelfPublicKey := ihgs.computeShardForSelfPublicKey(nodesConfig)
-	nodesConfig.shardID = shardIDForSelfPublicKey
+	log.Warn("Computed shardId", "shardId", nodesConfig.shardID)
```
Comment (SebastianMarian) [inline sharding/indexHashedNodesCoordinator.go:220]: Could be set to Debug level or removed if it is not really necessary

File: storage/factory/openStorage.go [hunk 9]
```diff
@@ -117,6 +118,20 @@ func (o *openStorageUnits) OpenStorageUnits() (storage.Storer, error) {
...
+func createDB(persisterFactory *PersisterFactory, persisterPath string) (storage.Persister, error) {
+	var persister storage.Persister
+	var err error
+	for i := 0; i < core.MaxRetriesToCreateDB; i++ {
+		persister, err = persisterFactory.Create(persisterPath)
+		if err == nil {
+			return persister, nil
+		}
+		log.Warn("Create Persister failed", "path", persisterPath)
+		time.Sleep(core.SleepTimeBetweenCreateDBRetries * time.Second)
```
Comment (SebastianMarian) [inline storage/factory/openStorage.go:130]: time.Sleep(core.SleepTimeBetweenCreateDBRetries)

File: core/constants.go [hunk 10]
```diff
@@ -338,3 +338,12 @@ const (
...
+const SleepTimeBetweenCreateDBRetries = 5
```
Comment (SebastianMarian) [inline core/constants.go:349]: 5 * time.Second

File: scripts/testnet/variables.sh (hunk 5 above)
Comment (raduchis) [reply scripts/testnet/variables.sh:52]: nope, updone

File: sharding/indexHashedNodesCoordinator.go (hunk 6 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:120]: fixed

File: sharding/indexHashedNodesCoordinator.go (hunk 7 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:213]: no. removed

File: sharding/indexHashedNodesCoordinator.go (hunk 8 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:220]: remove

File: cmd/node/main.go [hunk 11]
```diff
@@ -322,6 +324,12 @@ VERSION:
 		Value: uint64(2),
 	}
 
+	startInEpoch = cli.BoolFlag{
+		Name: "start-in-epoch",
+		Usage: "Boolean option for enabling a node the fast bootstrap mechanism from the network is enabled if " +
```
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/node/main.go:329]: fixed

File: cmd/node/main.go [hunk 12]
```diff
@@ -798,8 +814,12 @@ func startNode(ctx *cli.Context, log logger.Logger, version string) error {
 	if ctx.IsSet(numActivePersisters.Name) {
 		generalConfig.StoragePruning.NumActivePersisters = ctx.GlobalUint64(numActivePersisters.Name)
 	}
-
-	chanStopNodeProcess := make(chan bool, 1)
+	log.Info("Bootstrap", "epoch", bootstrapParameters.Epoch)
+	log.Info("nodesConfig is", "not nil", bootstrapParameters.NodesConfig != nil)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/node/main.go:822]: fixed

File: cmd/node/main.go [hunk 13]
```diff
@@ -1089,11 +1111,12 @@ func startNode(ctx *cli.Context, log logger.Logger, version string) error {
...
 	sigs := make(chan os.Signal, 1)
 	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	var sig endProcess.EndProcessArgument
 	select {
 	case <-sigs:
 		log.Info("terminating at user's signal...")
-	case <-chanStopNodeProcess:
-		log.Info("terminating at internal stop signal...")
+	case sig = <-chanStopNodeProcess:
+		log.Info(fmt.Sprintf("terminating at internal stop signal... reason: %s", sig.Reason))
```
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/node/main.go:1124]: fixed

File: cmd/node/main.go (hunk 1 above)
Comment (raduchis) [reply cmd/node/main.go:1152]: fxied

File: cmd/node/main.go (hunk 2 above)
Comment (raduchis) [reply cmd/node/main.go:1172]: fixed

File: cmd/node/main.go (hunk 3 above)
Comment (raduchis) [reply cmd/node/main.go:1566]: fixed

File: core/constants.go (hunk 4 above)
Comment (raduchis) [reply core/constants.go:351]: fixed

File: storage/factory/openStorage.go (hunk 9 above)
Comment (raduchis) [reply storage/factory/openStorage.go:130]: fixed

File: core/constants.go (hunk 10 above)
Comment (raduchis) [reply core/constants.go:352]: fixed

File: cmd/node/main.go [hunk 14]
```diff
@@ -346,6 +354,8 @@ func main() {
 	_ = logger.SetDisplayByteSlice(logger.ToHexShort)
 	log := logger.GetOrCreate("main")
 
+	time.Sleep(2 * time.Second)
```
Comment (sasurobert) [inline cmd/node/main.go:357]: why is this needed here ? add a constant.

File: cmd/node/main.go [hunk 15]
```diff
@@ -1121,9 +1144,49 @@ func startNode(ctx *cli.Context, log logger.Logger, version string) error {
...
+func handleAppClose(log logger.Logger, endProcessArgument endProcess.EndProcessArgument) {
+	log.Debug(
+		"restarting node",
+		"reason",
+		endProcessArgument.Reason,
+		"description",
+		endProcessArgument.Description)
+	switch endProcessArgument.Reason {
+	case core.ShuffledOut:
+		{
```
Comment (sasurobert) [inline cmd/node/main.go:1161]: identation

File: cmd/node/main.go [hunk 16]
```diff
@@ -1481,6 +1562,35 @@ func createNodesCoordinator(
 	return nodesCoordinator, nil
 }
 
+func serializableValidatorsToValidators(nodeRegistryValidators map[string][]*sharding.SerializableValidator) (map[uint32][]sharding.Validator, error) {
```
Comment (sasurobert) [inline cmd/node/main.go:1565]: make the methods in the sharding folder public - no need in main

File: cmd/node/main.go [hunk 17]
```diff
@@ -1481,6 +1562,35 @@ func createNodesCoordinator(
...
+		if err != nil {
+			return nil, err
+		}
+		validators[uint32(shardIdInt)] = newValidators
+	}
+
+	return validators, nil
+}
+
+func serializableShardValidatorListToValidatorList(shardValidators []*sharding.SerializableValidator) ([]sharding.Validator, error) {
```
Comment (sasurobert) [inline cmd/node/main.go:1582]: make the methods in the sharding folder public - no need in main

File: data/endProcess/endProcessArgument.go [hunk 18]
```diff
@@ -0,0 +1,6 @@
+package endProcess
+
+type EndProcessArgument struct {
```
Comment (sasurobert) [inline data/endProcess/endProcessArgument.go:3]: add comment

File: sharding/indexHashedNodesCoordinator.go [hunk 19]
```diff
@@ -134,7 +134,16 @@ func NewIndexHashedNodesCoordinator(arguments ArgNodesCoordinator) (*indexHashed
 		log.Error("saving initial nodes coordinator config failed",
 			"error", err.Error())
 	}
-
+	log.Info("new nodes config is set for epoch", "epoch", arguments.Epoch)
+	currentNodesConfig := ihgs.nodesConfig[arguments.Epoch]
+	if currentNodesConfig != nil {
```
Comment (sasurobert) [inline sharding/indexHashedNodesCoordinator.go:139]: if no nodesConfig - shouldn't we return error ?

File: sharding/indexHashedNodesCoordinator.go [hunk 20]
```diff
@@ -623,7 +631,14 @@ func (ihgs *indexHashedNodesCoordinator) EpochStartAction(hdr data.HeaderHandler
 			}
 		}
 	}
+
+	nodesConfig := ihgs.nodesConfig[newEpoch]
 	ihgs.mutNodesConfig.Unlock()
+
+	err = ihgs.shuffledOutHandler.Process(nodesConfig.shardID)
```
Comment (sasurobert) [inline sharding/indexHashedNodesCoordinator.go:638]: we might think about calling shuffle out only when the epoch start shard header gets finalized.

File: cmd/node/main.go (hunk 14 above)
Comment (raduchis) [reply cmd/node/main.go:357]: fixed. moved to the startInEpoch flag verification.

File: cmd/node/main.go (hunk 16 above)
Comment (raduchis) [reply cmd/node/main.go:1565]: fixed

File: cmd/node/main.go (hunk 17 above)
Comment (raduchis) [reply cmd/node/main.go:1582]: fixed

File: sharding/indexHashedNodesCoordinator.go (hunk 19 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:139]: fixed. Should not happen, because the nodesConfig[currentEpoch] was set just above

File: sharding/indexHashedNodesCoordinator.go (hunk 20 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:638]: this happens on the SetProcessed of the metachain in the shard. Aren't the headers finalized when processing them in a different shard?

Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:642]: added todo

File: process/block/shardblock.go [hunk 21]
```diff
@@ -915,10 +913,14 @@ func (sp *shardProcessor) updateState(headers []data.HeaderHandler, currentHeade
 			log.Debug("could not get shard header from storage")
 			return
 		}
+		currentHeader := headers[i]
```
Comment (sasurobert) [inline process/block/shardblock.go:916]: may change for i := range headers to for _, hdr := range headers then this line is not more needed.
